In [86]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn import svm

In [79]:
df = pd.read_csv('train.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [80]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 889 entries, 0 to 888
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  889 non-null    int64  
 1   Survived     889 non-null    int64  
 2   Pclass       889 non-null    int64  
 3   Name         889 non-null    object 
 4   Sex          889 non-null    object 
 5   Age          889 non-null    float64
 6   SibSp        889 non-null    int64  
 7   Parch        889 non-null    int64  
 8   Ticket       889 non-null    object 
 9   Fare         889 non-null    float64
 10  Cabin        202 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.5+ KB


In [81]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

In [82]:
df.drop(['PassengerId', 'Name', 'Ticket', 'Cabin'], axis=1, inplace=True)

df.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

In [83]:
# Using LabelEncoder on 'Sex' & 'Embarked'

le = LabelEncoder()

In [84]:
df['Sex'] = le.fit_transform(df['Sex'])
df['Sex'].value_counts()

1    577
0    312
Name: Sex, dtype: int64

In [85]:
df['Embarked'] = le.fit_transform(df['Embarked'])
df['Embarked'].value_counts()

2    644
0    168
1     77
Name: Embarked, dtype: int64

In [94]:
# Split data into X and y for IDV and DV.

X = df.drop(['Pclass'], axis=1)
y = df['Pclass']

In [95]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [96]:
model = svm.SVC(gamma=0.01, C=100)

In [97]:
model.fit(X_train, y_train)

SVC(C=100, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.01, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [98]:
y_pred = model.predict(X_test)

In [99]:
accuracy_score(y_test, y_pred, normalize=True)

0.9026217228464419

In [100]:
confusion_matrix(y_test, y_pred)

array([[ 64,   4,   2],
       [  4,  36,   9],
       [  1,   6, 141]], dtype=int64)

A total observations of 267 are recorded of which 241 are correct and the rest incorrect.

In [101]:
df.columns

Index(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare',
       'Embarked'],
      dtype='object')

In [93]:
def svm_model():
    features = ['Survived', 'Pclass', 'Sex', 'Parch', 'SibSp', 'Embarked']
    
    for i in features:
        X = df.drop(i, axis=1)
        y = df[i]
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
        model = svm.SVC(gamma=0.01, C=100)
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)
        print('The Dependent Variable is', i, 'and the rest are the Independent Variable')
        print('The Accuracy Score is : ', accuracy_score(y_test, y_pred, normalize=True))
        print('The corresponding Confusion Matrix is :\n', confusion_matrix(y_test, y_pred))
        print('\n')
    
svm_model()

The Dependent Variable is Survived and the rest are the Independent Variable
The Accuracy Score is :  0.7415730337078652
The corresponding Confusion Matrix is :
 [[124  33]
 [ 36  74]]


The Dependent Variable is Pclass and the rest are the Independent Variable
The Accuracy Score is :  0.9026217228464419
The corresponding Confusion Matrix is :
 [[ 64   4   2]
 [  4  36   9]
 [  1   6 141]]


The Dependent Variable is Sex and the rest are the Independent Variable
The Accuracy Score is :  0.7265917602996255
The corresponding Confusion Matrix is :
 [[ 66  32]
 [ 41 128]]


The Dependent Variable is Parch and the rest are the Independent Variable
The Accuracy Score is :  0.797752808988764
The corresponding Confusion Matrix is :
 [[187   6   5   0   0   0]
 [ 21  15   3   0   0   1]
 [ 10   6  11   0   0   0]
 [  0   0   1   0   0   0]
 [  1   0   0   0   0   0]
 [  0   0   0   0   0   0]]


The Dependent Variable is SibSp and the rest are the Independent Variable
The Accuracy Score is :  0

The Pclass records the highest accuracy score of 90.26%